File: src/main/java/com/scottscmo/ppt/TemplateHandler.java
package com.scottscmo.ppt;

import org.apache.poi.xslf.usermodel.*;
import org.openxmlformats.schemas.drawingml.x2006.main.CTTextParagraph;

import java.util.List;
import java.util.Map;

public interface TemplateHandler {
    void evaluateTemplate(XMLSlideShow ppt, int index);

    default void replaceText(XSLFSlide slide, Map<String, String> replacements) {
        for (XSLFTextShape textShape : slide.getPlaceholders()) {
            String text = textShape.getText();
            for (String searchText : replacements.keySet()) {
                if (text.contains(searchText)) {
                    text = text.replace(searchText, replacements.get(searchText));
                }
            }
            clearText(textShape);
            appendText(textShape, text);
        }
    }

    default void appendText(XSLFTextShape textShape, String text) {
        String[] lines = text.trim().split("\n");
        List<XSLFTextParagraph> pps = textShape.getTextParagraphs();
        XSLFTextParagraph pp = pps.get(pps.size() - 1);

        for (int i = 0; i < lines.length; i++) {
            pp.addNewTextRun().setText(lines[i]);
            if (i + 1 < lines.length) {
                pp.addLineBreak();
            }
        }
    }

    /**
     * Helper to remove all texts and new lines.
     * XSLFTextShape.setText("") is bugged when there's new line
     */
    default void clearText(XSLFTextShape textShape) {
        for (XSLFTextParagraph pp : textShape.getTextParagraphs()) {
            for (XSLFTextRun textRun : pp.getTextRuns()) {
                if (!textRun.getRawText().equals("\n")) {
                    textRun.setText("");
                }
            }
            CTTextParagraph ppxml = pp.getXmlObject();
            for (int i = ppxml.sizeOfBrArray(); i>0; i--) {
                ppxml.removeBr(i-1);
            }
        }
    }

    default void replaceText(XSLFSlide slide, String searchText, String replacement) {
        replaceText(slide, Map.of(searchText, replacement));
    }

    default String findText(XSLFSlide slide, String searchText) {
        for (XSLFTextShape textShape : slide.getPlaceholders()) {
            String text = textShape.getText();
            if (text.contains(searchText)) {
                return text;
            }
        }
        return null;
    }

    default XSLFSlideMaster getSlideMaster(XMLSlideShow ppt, String name) {
        if (ppt == null || name == null || name.isEmpty()) return null;

        return ppt.getSlideMasters().stream()
                .filter(master -> master.getTheme().getName().equals(name))
                .findFirst()
                .orElse(null);
    }
}

File: src/main/kotlin/com/scottscmo/ppt/TemplateHandler.kt
package com.scottscmo.ppt

import org.apache.poi.xslf.usermodel.*

interface TemplateHandler {
    fun evaluateTemplate(ppt: XMLSlideShow, index: Int) {

    }

    fun replaceText(slide: XSLFSlide, replacements: Map<String, String>) {
        for (textShape in slide.placeholders) {
            var text = textShape.text
            for ((searchText, replacement) in replacements) {
                if (text.contains(searchText)) {
                    text = text.replace(searchText, replacement)
                }
            }
            clearText(textShape)
            appendText(textShape, text)
        }
    }

    fun appendText(textShape: XSLFTextShape, text: String) {
        val lines = text.trim().split("\n")
        val pps = textShape.textParagraphs
        val pp = pps[pps.size - 1]
        for (i in lines.indices) {
            pp.addNewTextRun().setText(lines[i])
            if (i + 1 < lines.size) {
                pp.addLineBreak()
            }
        }
    }

    /**
     * Helper to remove all texts and new lines.
     * XSLFTextShape.setText("") is bugged when there's new line
     */
    fun clearText(textShape: XSLFTextShape) {
        for (pp in textShape.textParagraphs) {
            for (textRun in pp.textRuns) {
                if (textRun.rawText != "\n") {
                    textRun.setText("")
                }
            }
            val ppxml = pp.xmlObject
            for (i in ppxml.sizeOfBrArray() downTo 1) {
                ppxml.removeBr(i - 1)
            }
        }
    }

    fun replaceText(slide: XSLFSlide, searchText: String, replacement: String) {
        replaceText(slide, mapOf(searchText to replacement))
    }

    fun findText(slide: XSLFSlide, searchText: String): String? {
        for (textShape in slide.placeholders) {
            val text = textShape.text
            if (text.contains(searchText)) {
                return text
            }
        }
        return null
    }

    fun getSlideMaster(ppt: XMLSlideShow, name: String): XSLFSlideMaster? {
        return ppt.slideMasters.firstOrNull { it.theme.name == name }
    }

    fun getSlideMasterLayout(ppt: XMLSlideShow, name: String, layout: String): XSLFSlideLayout? {
        return getSlideMaster(ppt, name)?.getLayout(layout)
    }
}